doc/design/queue_api.md

```diff
@@ -0,0 +1,382 @@
+* Each queue must meet `Allocated >= Used`, it means the queue won't be overused
+
+####Brief workflow:
+
+* Preprocess of `QueueInfo`
+* Adjust Deserved/Allocated/Used/Preempting of each Queue to trigger preemption, then update results to each Queue
+* Terminate running pods for each Queue which need preemption, and update Allocated of Queue after preempted pod is finished
+
+```go
+type Interface interface {
+	// Run start pod informer to handle terminating pods
+	Run(stopCh <-chan struct{})
+
+	// Preprocessing preprocess for each queue
+	// Currently, it terminates pods of overused(Allocated < Used) queue without preemption
+	// Deserved/Allocated/Used/Preempting will not be adjusted
+	Preprocessing(queues map[string]*schedulercache.QueueInfo, pods []*schedulercache.PodInfo) (map[string]*schedulercache.QueueInfo, error)
+
+	// PreemptResources preempt resources between queue
+	// Deserved/Allocated/Used/Preempting will be adjusted
+	PreemptResources(queues map[string]*schedulercache.QueueInfo) error
+}
+```
+
+####Details of preemption
+
+#####Preprocess stage
+Currently, this stage terminates pods of the queue which `Allocated < Used` to avoid overuse(make queue `Allocated >= Used`), preemption will not be triggered between queues.
+
+```
+For examples:
+--------------------------------------------------------------------------
+| Queue-1                                                                |
+|    Weight: 2                                                           |
+|                                                                        |
+|    Allocated:        Used:                Pods:                        |
+|      cpu: 3            cpu: 5               pod-1: cpu=2 memory=1Gi    |
+|      memory: 9Gi       memory: 3Gi          pod-2: cpu=2 memory=1Gi    |
+|                                             pod-3: cpu=1 memory=1Gi    |
+--------------------------------------------------------------------------
+
+--------------------------------------------------------------------------
+| Queue-2                                                                |
+|    Weight: 4                                                           |
+|                                                                        |
+|    Allocated:        Used:                Pods:                        |
+|      cpu: 6            cpu: 5               pod-1: cpu=2 memory=1Gi    |
+|      memory: 18Gi      memory: 3Gi          pod-2: cpu=2 memory=1Gi    |
+|                                             pod-3: cpu=1 memory=1Gi    |
+--------------------------------------------------------------------------
+
+There are 9 CPUs totally. 3 CPUs are allocated to Queue-1 and 6 CPUs are allocated to Queue-2.
+However, Queue-1 is overused due to some race condition. So pod-1 will be chosen to terminate. The pod is randomly selected currently.
+--------------------------------------------------------------------------
```

Comment:
Is it possible to provide a callback function to choose Pods to be killed?

Reply:
yes, we can enhance the pod selection, such as provide some default selection functions or a callback.
Currently, the pod is chosen randomly.